### PR TITLE
feat(enhancement): Hide unknown system names in the escort HUD

### PR DESCRIPTION
**Pull request by TomGoodIdea**
## Feature Details
~~Sometimes it's faster for an escort to travel through visible but unexplored systems (or some hidden ones, but that's another problem). Since escorts aren't able to map systems they visit, it's unreasonable to know the names as the player won't even know where the systems are.~~ Edit: That's fixed, but an escort may happen to be in an unexplored system as a result of "unvisit" events.

In such case the system name in the escort HUD and the Player Info panel is replaced with "???". However, sorting the fleet list by system name still takes true names (see the comments below).

## Testing Done
Yes.

## Performance Impact
N/A

### CODE REVIEW

**Comment by lumbar527** [comment]
Your escorts might give you feedback or something. Actually, your escorts should give you map data.

**Comment by DarcyManoel** [comment]
> Your escorts might give you feedback or something. Actually, your escorts should give you map data.

Would it logically follow that escorts travelling through new systems, should reveal those systems for the player?

**Comment by lumbar527** [comment]
I think I meant that when I said
> Actually, your escorts should give you map data.

**Comment by Zitchas** [comment]
Really, your escorts should never travel through unknown systems. After all, they have no way of knowing that it might be faster to do so. For all they know they might be dead ends and get stuck. So the best solution would be to ensure that escorts are forcibly restricted to only using systems the player has mapped, and if there is no way for them to reach the player, then they stay wherever they are.

From a more meta perspective, I think having our escorts starting to do exploration to try to figure out routes to the player has a few problems. Most of these already exist, just having escorts explore their way to the player makes them more obvious.
a) Makes it hard for the player to tell where they have explored and where they haven't.
b) Makes the current problems of time passage even more problematic. (it takes us a day to go one system... Why can my escorts explore dozens of systems on their way to finding a way to me in less than a day?)
c) Results in non-sensical story elements, such as NPCs that talk to the player immediately when they enter the system. Why didn't they talk to the escorts? Or if they did, why didn't they report it to you?
d) If a system is too dangerous for the player to get through, why can the player's escort freighter-with-no-weapons get through just fine?
e) Escorts are theoretically supposed to be just a balance element to help the player with challenges. A way to supplement their flagship. They aren't supposed to be independently operating secondary fleets. Having escorts exploring makes them more like NPCs and less like members of the player's fleet, which pushes the game into more of a 4x feel. Is this a direction we want to go?

**Comment by TomGoodIdea** [comment]
> Really, your escorts should never travel through unknown systems.

#6127, but its current state is unclear to me

**Comment by warp-core** [comment]
I think it might be better to use "Unexplored System," which is how systems the player has not visited are referred to in other places (map, mini-map, destination name in the HUD).

**Comment by TomGoodIdea** [comment]
Screenshot_20230602_130539
It seems a bit too long.

Maybe we could make too long names display with a "..." at the end ("Unknown Sys..." in this case) and hovering would reveal the full name.

**Comment by TomGoodIdea** [comment]
Screenshot_20230602_131604
It seems a bit too long.

But maybe we could display too long names with "..." at the end ("Unexplored S..." in this case) and hovering would reveal the full name.

**Comment by Zireael07** [comment]
I'd prefer ??? or "Unexplored" if the full "Unexplored System" doesn't fit.

**Comment by quyykk** [comment]
> https://github.com/endless-sky/endless-sky/pull/6127, but its current state is unclear to me

I'll finish it :smile:

**Comment by TomGoodIdea** [comment]
With #6127 reopened I think this change is unnecessary.

**Comment by quyykk** [comment]
I will put in my vote for "???" since it seems to be more clear at a glance that the ship is in an unknown system. just imo

File: source/EscortDisplay.cpp [hunk 1]
```diff
@@ -202,7 +202,7 @@ EscortDisplay::Icon::Icon(const Ship &ship, bool isHere, bool fleetIsJumping, bo
 	cannotJump(fleetIsJumping && !ship.IsHyperspacing() && !ship.JumpsRemaining()),
 	isSelected(isSelected),
 	cost(ship.Cost()),
-	system((!isHere && ship.GetSystem()) ? ship.GetSystem()->Name() : ""),
+	system((!isHere && ship.GetSystem()) ? systemNameKnown ? ship.GetSystem()->Name() : "Unexplored System" : ""),
```
Comment (Amazinite) [inline source/EscortDisplay.cpp:205]:
```suggestion
	system((!isHere && ship.GetSystem()) ? (systemNameKnown ? ship.GetSystem()->Name() : "???") : ""),
```

File: source/Engine.cpp [hunk 2]
```diff
@@ -621,7 +621,8 @@ void Engine::Step(bool isActive)
 			if(!it->IsYours() && !it->CanBeCarried())
 			{
 				bool isSelected = (flagship && flagship->GetTargetShip() == it);
-				escorts.Add(*it, it->GetSystem() == currentSystem, fleetIsJumping, isSelected);
+				const System *sys = it->GetSystem();
+				escorts.Add(*it, sys == currentSystem, player.KnowsName(*sys), fleetIsJumping, isSelected);
```
Comment (Amazinite) [inline source/Engine.cpp:625]:
```suggestion
				const System *system = it->GetSystem();
				escorts.Add(*it, system == currentSystem, player.KnowsName(*system), fleetIsJumping, isSelected);
```

File: source/Engine.cpp [hunk 3]
```diff
@@ -634,7 +635,8 @@ void Engine::Step(bool isActive)
 					isSelected = true;
 					break;
 				}
-			escorts.Add(*escort, escort->GetSystem() == currentSystem, fleetIsJumping, isSelected);
+			const System *sys = escort->GetSystem();
+			escorts.Add(*escort, sys == currentSystem, player.KnowsName(*sys), fleetIsJumping, isSelected);
```
Comment (Amazinite) [inline source/Engine.cpp:639]:
```suggestion
			const System *system = escort->GetSystem();
			escorts.Add(*escort, system == currentSystem, player.KnowsName(*system), fleetIsJumping, isSelected);
```

**Review by Amazinite — COMMENTED** [review]
We should probably make this change in PlayerInfoPanel as well.
https://github.com/endless-sky/endless-sky/blob/a2d47c4457e0c491decc0521424cbddf1614dc92/source/PlayerInfoPanel.cpp#L773-L774

**Comment by TomGoodIdea** [comment]
I need #8529 merged to make the change in Player Info panel, because there's the system name comparator function that has to be static and can't use PlayerInfoPanel's player.

**Comment by quyykk** [comment]
Alternatively, we can keep the sort by real system name even though the system names in the playerinfopanel themselves are shown as ???. I think that's fine tbh. Thoughts?

**Comment by TomGoodIdea** [comment]
hmm, sounds acceptable for now, we can change that later

File: source/PlayerInfoPanel.cpp [hunk 4]
```diff
@@ -771,7 +771,7 @@ void PlayerInfoPanel::DrawFleet(const Rectangle &bounds)
 		table.Draw(ship.DisplayModelName());
 
 		const System *system = ship.GetSystem();
-		table.Draw(system ? system->Name() : "");
+		table.Draw(system ? (player.KnowsName(*system) ? ship.GetSystem()->Name() : "???") : "");
```
Comment (quyykk) [inline source/PlayerInfoPanel.cpp:774]:
```suggestion
		table.Draw(system ? (player.KnowsName(*system) ? system->Name() : "???") : "");
```

**Review by quyykk — APPROVED** [review]

**Pull request merged**